File: cprdspy/CPR_js2/Circles/circle.js
/**
 * CPR 图形库 - JavaScript 版本 (Plotly.js)
 * Circles 模块 - 圆形和椭圆相关函数
 */

/**
 * 绘制圆形
 * @param {Object} options - 配置选项
 * @returns {Object} Plotly 轨迹对象
 */
function circle({
    radius = 1,
    color = "#0F0",
    alpha = 1,
    center = [0, 0],
    points = 100,
    lineWidth = 1,
    label = null,
    name = null,
} = {}) {
    const angle = Array.from({ length: points + 1 }, (_, i) => (2 * Math.PI * i) / points);
    const x = angle.map(a => center[0] + radius * Math.cos(a));
    const y = angle.map(a => center[1] + radius * Math.sin(a));

    return {
        type: "scattergl",
        mode: "lines",
        x: x,
        y: y,
        line: { color: color, width: lineWidth },
        opacity: alpha,
        name: name || label,
        hoverinfo: "none",
    };
}

/**
 * 通过圆心和圆上一点绘制圆
 */
function circleP(center, point, color = "b") {
    const dx = point[0] - center[0];
    const dy = point[1] - center[1];
    const radius = Math.sqrt(dx * dx + dy * dy);
    return circle({ radius, color, center });
}

/**
 * 绘制椭圆
 */
function ellipse({
    a = 2,
    b = 1,
    rotation = 0,
    color = "#0f0",
    alpha = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    label = null,
} = {}) {
    const angleRad = (rotation * Math.PI) / 180;
    const theta = Array.from({ length: points + 1 }, (_, i) => (2 * Math.PI * i) / points);

    const x = theta.map(t =>
        a * Math.cos(t) * Math.cos(angleRad) - b * Math.sin(t) * Math.sin(angleRad) + center[0]
    );
    const y = theta.map(t =>
        a * Math.cos(t) * Math.sin(angleRad) + b * Math.sin(t) * Math.cos(angleRad) + center[1]
    );

    return {
        type: "scattergl",
        mode: "lines",
        x: x,
        y: y,
        line: { color: color, width: lineWidth },
        opacity: alpha,
        name: label,
        hoverinfo: "none",
    };
}

/**
 * 绘制同心圆
 */
function concentricCircles({
    center = [0, 0],
    n = 3,
    radius = 1,
    param = 0.5,
    mode = "arithmetic",
    direction = "both",
    color = "#0F0",
    alpha = 1,
    points = 200,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const traces = [];

    if (showCenter) {
        traces.push({
            type: "scattergl",
            mode: "markers",
            x: [center[0]],
            y: [center[1]],
            marker: { color: color, size: 6 },
            showlegend: false,
            hoverinfo: "none",
        });
    }

    const radii = [];
    const d = parseFloat(param);
    const q = parseFloat(param);

    if (mode.toLowerCase().startsWith("a")) {
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius + i * d);
                radii.push(radius - i * d);
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) radii.push(radius + i * d);
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) radii.push(radius - i * d);
        }
    } else if (mode.toLowerCase().startsWith("g")) {
        if (q === 0) throw new Error("ratio (param) must be non-zero for geometric mode");
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius * Math.pow(q, i));
                radii.push(radius / Math.pow(q, i));
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) radii.push(radius * Math.pow(q, i));
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) radii.push(radius / Math.pow(q, i));
        }
    }

    radii.push(radius);
    const uniqueRadii = [...new Set(radii)].filter(r => r > 0).sort((a, b) => a - b);

    for (const r of uniqueRadii) {
        const angle = Array.from({ length: points + 1 }, (_, i) => (2 * Math.PI * i) / points);
        const x = angle.map(a => center[0] + r * Math.cos(a));
        const y = angle.map(a => center[1] + r * Math.sin(a));

        traces.push({
            type: "scattergl",
            mode: "lines",
            x: x,
            y: y,
            line: { color: color, width: lineWidth },
            opacity: alpha,
            showlegend: false,
            hoverinfo: "none",
        });
    }

    return traces;
}

/**
 * 绘制同心椭圆
 */
function concentricEllipses({
    center = [0, 0],
    n = 3,
    a = 2,
    b = 1,
    param = 0.5,
    mode = "arithmetic",
    direction = "both",
    rotation = 0,
    color = "#0F0",
    alpha = 1,
    points = 200,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const traces = [];

    if (showCenter) {
        traces.push({
            type: "scattergl",
            mode: "markers",
            x: [center[0]],
            y: [center[1]],
            marker: { color: color, size: 6 },
            showlegend: false,
            hoverinfo: "none",
        });
    }

    const ellipses = [];
    const d = parseFloat(param);
    const q = parseFloat(param);

    if (mode.toLowerCase().startsWith("a")) {
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a + i * d, b + i * d]);
                ellipses.push([a - i * d, b - i * d]);
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) ellipses.push([a + i * d, b + i * d]);
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) ellipses.push([a - i * d, b - i * d]);
        }
    } else if (mode.toLowerCase().startsWith("g")) {
        if (q === 0) throw new Error("ratio (param) must be non-zero for geometric mode");
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a * Math.pow(q, i), b * Math.pow(q, i)]);
                ellipses.push([a / Math.pow(q, i), b / Math.pow(q, i)]);
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) ellipses.push([a * Math.pow(q, i), b * Math.pow(q, i)]);
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) ellipses.push([a / Math.pow(q, i), b / Math.pow(q, i)]);
        }
    }

    ellipses.push([a, b]);
    const uniqueMap = new Map();
    for (const [aa, bb] of ellipses) {
        if (aa > 0 && bb > 0) {
            uniqueMap.set(`${aa.toFixed(12)},${bb.toFixed(12)}`, [aa, bb]);
        }
    }
    const sortedEllipses = [...uniqueMap.values()].sort(([a1, b1], [a2, b2]) => a1 * b1 - a2 * b2);
    const angleRad = (rotation * Math.PI) / 180;

    for (const [aa, bb] of sortedEllipses) {
        const theta = Array.from({ length: points + 1 }, (_, i) => (2 * Math.PI * i) / points);
        const x = theta.map(t =>
            aa * Math.cos(t) * Math.cos(angleRad) - bb * Math.sin(t) * Math.sin(angleRad) + center[0]
        );
        const y = theta.map(t =>
            aa * Math.cos(t) * Math.sin(angleRad) + bb * Math.sin(t) * Math.cos(angleRad) + center[1]
        );

        traces.push({
            type: "scattergl",
            mode: "lines",
            x: x,
            y: y,
            line: { color: color, width: lineWidth },
            opacity: alpha,
            showlegend: false,
            hoverinfo: "none",
        });
    }

    return traces;
}

export { circle, circleP, ellipse, concentricCircles, concentricEllipses };

File: cprdspy/CPR_plotly1/Circles/circle.py
import numpy as np
import plotly.graph_objects as go


def circle(
    radius=1,
    color="#0F0",
    alpha=1,
    center=(0, 0),
    points=100,
    line_width=1,
    label=None,
    mode="lines",
    name=None,
    showlegend=None,
    **kwargs,
):
    """
    绘制一个圆。

    参数:
        radius (float): 圆的半径，默认为 1。
        color (str): 圆的颜色，默认为绿色 ("#0F0")。
        alpha (float): 透明度，默认为 1（不透明）。
        center (tuple): 圆心坐标，默认为 (0, 0)。
        points (int): 用于绘制圆的点数，默认为 100。
        line_width (float): 线宽，默认为 1。
        label (str): 图例标签，默认为 None。
        mode (str): 绘图模式，默认为 "lines"。
        name (str): 轨迹名称，默认为 None。
        showlegend (bool): 是否显示图例，默认为 None。
        **kwargs: 其他传递给 go.Scatter 的参数。

    返回:
        plotly.graph_objs.Scatter: 圆的轨迹对象。
    """
    angle = np.linspace(0, 2 * np.pi, points)
    x = center[0] + radius * np.cos(angle)
    y = center[1] + radius * np.sin(angle)

    trace = go.Scatter(
        x=x,
        y=y,
        mode=mode,
        line=dict(color=color, width=line_width, dash="solid"),
        fill="none",
        opacity=alpha,
        name=name if name else label,
        showlegend=showlegend,
        hoverinfo="none" if not label else "x+y+text",
        text=label,
        **kwargs,
    )

    return trace


def draw_circle(
    radius=1,
    color="#0F0",
    alpha=1,
    center=(0, 0),
    points=100,
    line_width=1,
    label=None,
):
    """
    绘制圆并返回 Figure（便捷函数）。
    """
    trace = circle(
        radius=radius,
        color=color,
        alpha=alpha,
        center=center,
        points=points,
        line_width=line_width,
        label=label,
    )
    fig = go.Figure(data=[trace])
    fig.update_layout(
        xaxis=dict(scaleanchor="y", scaleratio=1),
        yaxis=dict(scaleanchor="x", scaleratio=1),
        width=600,
        height=600,
    )
    return fig


def circle_p(center, point, color="b"):
    """
    通过圆心和圆上一点绘制圆。
    """
    radius = np.sqrt((point[0] - center[0]) ** 2 + (point[1] - center[1]) ** 2)
    return circle(radius=radius, color=color, center=center)


# 椭圆
def ellipse(
    a=2,
    b=1,
    rotation=0,
    color="#0f0",
    alpha=1,
    center=(0, 0),
    points=1000,
    line_width=1,
    label=None,
):
    """
    绘制椭圆。

    参数:
        a (float): 半长轴。
        b (float): 半短轴。
        rotation (float): 旋转角度（度）。
        color (str): 颜色。
        alpha (float): 透明度。
        center (tuple): 中心坐标。
        points (int): 点数。
        line_width (float): 线宽。
        label (str): 标签。

    返回:
        plotly.graph_objs.Scatter: 椭圆轨迹。
    """
    theta = np.linspace(0, 2 * np.pi, points)
    angle_rad = np.deg2rad(rotation)
    x = (
        a * np.cos(theta) * np.cos(angle_rad)
        - b * np.sin(theta) * np.sin(angle_rad)
        + center[0]
    )
    y = (
        a * np.cos(theta) * np.sin(angle_rad)
        + b * np.sin(theta) * np.cos(angle_rad)
        + center[1]
    )

    trace = go.Scatter(
        x=x,
        y=y,
        mode="lines",
        line=dict(color=color, width=line_width),
        opacity=alpha,
        name=label,
        showlegend=label is not None,
    )
    return trace


# 统一的同心圆函数
def concentric_circles(
    center=(0, 0),
    n=3,
    radius=1,
    param=0.5,
    mode="arithmetic",
    direction="both",
    color="#0F0",
    alpha=1,
    points=200,
    line_width=1,
    show_center=True,
    **kwargs,
):
    """
    绘制同心圆的统一函数（兼容等差与等比序列）。

    参数:
        center (tuple): 圆心坐标。
        n (int): 每一侧的圈数（当 direction=='both' 时每侧各 n 个）。
        radius (float): 基准半径（绘制在中间）。
        param (float): 当 mode=='arithmetic' 时作为公差 d；当 mode=='geometric' 时作为公比 q。
        mode (str): 'arithmetic' 或 'geometric'，选择等差或等比。
        direction (str): 'both'|'out'|'in'，控制向外/向内/双向绘制。
        color (str): 颜色。
        alpha (float): 透明度。
        points (int): 点数。
        line_width (float): 线宽。
        show_center (bool): 是否绘制圆心标记。
        **kwargs: 其它传入 go.Scatter 的参数。

    返回:
        list: 包含所有轨迹的列表。
    """
    traces = []

    if show_center:
        center_trace = go.Scatter(
            x=[center[0]],
            y=[center[1]],
            mode="markers",
            marker=dict(color=color, size=6),
            showlegend=False,
        )
        traces.append(center_trace)

    radii = []
    mode = mode.lower()
    direction = direction.lower()

    if mode.startswith("a"):
        d = float(param)
        if direction == "both":
            for i in range(1, n + 1):
                radii.append(radius + i * d)
                radii.append(radius - i * d)
        elif direction == "out":
            for i in range(1, n + 1):
                radii.append(radius + i * d)
        elif direction == "in":
            for i in range(1, n + 1):
                radii.append(radius - i * d)
    elif mode.startswith("g"):
        q = float(param)
        if q == 0:
            raise ValueError("ratio (param) must be non-zero for geometric mode")
        if direction == "both":
            for i in range(1, n + 1):
                radii.append(radius * (q**i))
                radii.append(radius / (q**i))
        elif direction == "out":
            for i in range(1, n + 1):
                radii.append(radius * (q**i))
        elif direction == "in":
            for i in range(1, n + 1):
                radii.append(radius / (q**i))

    radii = [r for r in radii if r > 0]
    radii.append(radius)
    radii = sorted(set(radii))

    for r in radii:
        angle = np.linspace(0, 2 * np.pi, points)
        x = center[0] + r * np.cos(angle)
        y = center[1] + r * np.sin(angle)
        trace = go.Scatter(
            x=x,
            y=y,
            mode="lines",
            line=dict(color=color, width=line_width),
            opacity=alpha,
            showlegend=False,
            **kwargs,
        )
        traces.append(trace)

    return traces


def concentric_ellipse(
    center=(0, 0),
    n=3,
    a=2,
    b=1,
    param=0.5,
    mode="arithmetic",
    direction="both",
    rotation=0,
    color="#0F0",
    alpha=1,
    points=200,
    line_width=1,
    show_center=True,
    **kwargs,
):
    """
    绘制同心椭圆（兼容等差与等比序列）。

    参数:
        center (tuple): 椭圆中心。
        n (int): 每侧的椭圆数量。
        a (float): 基准半长轴。
        b (float): 基准半短轴。
        param (float): 等差时为公差 d；等比时为公比 q。
        mode (str): 'arithmetic' 或 'geometric'。
        direction (str): 'both'|'out'|'in'。
        rotation (float): 椭圆旋转角度（度）。
        color (str): 颜色。
        alpha (float): 透明度。
        points (int): 点数。
        line_width (float): 线宽。
        show_center (bool): 是否绘制中心标记。

    返回:
        list: 包含所有轨迹的列表。
    """
    traces = []

    if show_center:
        center_trace = go.Scatter(
            x=[center[0]],
            y=[center[1]],
            mode="markers",
            marker=dict(color=color, size=6),
            showlegend=False,
        )
        traces.append(center_trace)

    ellipses = []
    mode = mode.lower()
    direction = direction.lower()

    if mode.startswith("a"):
        d = float(param)
        if direction == "both":
            for i in range(1, n + 1):
                ellipses.append((a + i * d, b + i * d))
                ellipses.append((a - i * d, b - i * d))
        elif direction == "out":
            for i in range(1, n + 1):
                ellipses.append((a + i * d, b + i * d))
        elif direction == "in":
            for i in range(1, n + 1):
                ellipses.append((a - i * d, b - i * d))
    elif mode.startswith("g"):
        q = float(param)
        if q == 0:
            raise ValueError("ratio (param) must be non-zero for geometric mode")
        if direction == "both":
            for i in range(1, n + 1):
                ellipses.append((a * (q**i), b * (q**i)))
                ellipses.append((a / (q**i), b / (q**i)))
        elif direction == "out":
            for i in range(1, n + 1):
                ellipses.append((a * (q**i), b * (q**i)))
        elif direction == "in":
            for i in range(1, n + 1):
                ellipses.append((a / (q**i), b / (q**i)))

    ellipses.append((a, b))
    uniq = {}
    for aa, bb in ellipses:
        if aa > 0 and bb > 0:
            key = (round(float(aa), 12), round(float(bb), 12))
            uniq[key] = (float(aa), float(bb))

    sorted_ellipses = sorted(uniq.values(), key=lambda t: t[0] * t[1])

    angle_rad = np.deg2rad(rotation)

    for aa, bb in sorted_ellipses:
        theta = np.linspace(0, 2 * np.pi, points)
        x = (
            aa * np.cos(theta) * np.cos(angle_rad)
            - bb * np.sin(theta) * np.sin(angle_rad)
            + center[0]
        )
        y = (
            aa * np.cos(theta) * np.sin(angle_rad)
            + bb * np.sin(theta) * np.cos(angle_rad)
            + center[1]
        )
        trace = go.Scatter(
            x=x,
            y=y,
            mode="lines",
            line=dict(color=color, width=line_width),
            opacity=alpha,
            showlegend=False,
            **kwargs,
        )
        traces.append(trace)

    return traces
